meteor/server/api/heapSnapshot.ts

```diff
@@ -0,0 +1,69 @@
+import * as v8 from 'node:v8'
+import { Readable } from 'stream'
+import { Meteor } from 'meteor/meteor'
+import Koa from 'koa'
+import KoaRouter from '@koa/router'
+import { fixValidPath } from '../../lib/lib'
+import { stringifyError } from '@sofie-automation/shared-lib/dist/lib/stringifyError'
+import { logger } from '../logging'
+import { Settings } from '../../lib/Settings'
+import { Credentials } from '../security/lib/credentials'
+import { SystemWriteAccess } from '../security/system'
+import { sleep } from '@sofie-automation/corelib/dist/lib'
+
+async function retrieveHeapSnapshot(cred0: Credentials): Promise<Readable> {
+	if (Settings.enableUserAccounts) {
+		await SystemWriteAccess.coreSystem(cred0)
+	}
+	logger.warn('Taking heap snapshot, expect system to be unresponsive for a few seconds..')
+	await sleep(100) // Allow the logger to catch up before continuing..
+
+	const stream = v8.getHeapSnapshot()
+	return stream
+}
+
+export const heapSnapshotPrivateApiRouter = new KoaRouter()
+
+// Setup endpoints:
+async function handleKoaResponse(ctx: Koa.ParameterizedContext, snapshotFcn: () => Promise<Readable>) {
+	if (ctx.query.areYouSure !== 'yes') {
+		ctx.response.status = 403
+		ctx.response.body = '?areYouSure=yes'
+		return
+	}
+
+	try {
+		const stream = await snapshotFcn()
+
+		ctx.response.type = 'application/octet-stream'
+		ctx.response.attachment(fixValidPath(`sofie-heap-snapshot-${new Date().toISOString()}.heapsnapshot`))
+		ctx.response.status = 200
+		ctx.body = stream
+		// ctx.response.body = JSON.stringify(snapshot, null, 4)
+	} catch (e) {
+		ctx.response.type = 'text/plain'
+		ctx.response.status = e instanceof Meteor.Error && typeof e.error === 'number' ? e.error : 500
+		ctx.response.body = 'Error: ' + stringifyError(e)
+
+		if (ctx.response.status !== 404) {
+			logger.error(stringifyError(e))
+		}
+	}
+}
+
+// For backwards compatibility:
+if (!Settings.enableUserAccounts) {
+	// Retrieve heap snapshot:
+	heapSnapshotPrivateApiRouter.get('/retrieve', async (ctx) => {
+		return handleKoaResponse(ctx, async () => {
+			return retrieveHeapSnapshot({ userId: null })
+		})
+	})
+}
+
+// Retrieve heap snapshot:
+heapSnapshotPrivateApiRouter.get('/:token/retrieve', async (ctx) => {
```

Comment:
I'm not sure that this route should exist.

In an environment where security has been enabled, I think this provides risk of being an accidental backdoor and I'm not 100% sure how it fits into the auth flow we have defined.

So until we address the user accounts setup, I think it would be better to omit this entirely. But I don't feel that strongly, so feel free to ignore this :)

Reply:
This is analog to [how the snapshots are retrieved](https://github.com/nrkno/sofie-core/blob/fix/take-heap-snapshot/meteor/server/api/snapshot.ts#L771), so I think that it should be fine to add it like this for now and they should both be addressed  when security is eventually looked into.